casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    merges(echoContract(1), echoContract(2), Rho("Nil"))
+  }
 
-        //multiparent block joining block0 and block1 since they do not conflict
-        multiparentBlock <- nodes(0).addBlock(deploys(1))
-        _                <- nodes(1).receive()
+  // TODO: Peek rows/column
+  // Note this skips pairs that lead to infinite loops
+  it should "handle multi-parent blocks correctly when they operate on volatile produce/consume pairs" in effectTest {
+    // Sends (linear sends)
+    val S0 = Rho("@0!(0)")
+    val S1 = Rho("@0!(1)")
+    // Repeats (persistent sends)
+    val R0 = Rho("@0!!(0)")
+    val R1 = Rho("@0!!(1)")
+    // For-s (linear receives)
+    val F_ = Rho("for (_ <- @0) { 0 }")
+    val F0 = Rho("for (@0 <- @0) { 0 }")
+    val F1 = Rho("for (@1 <- @0) { 0 }")
+    // Contracts (persistent receives)
+    val C_ = Rho("contract @0(id) = { 0 }")
+    val C0 = Rho("contract @0(@0) = { 0 }")
+    val C1 = Rho("contract @0(@1) = { 0 }")
+    //FIXME add missing cases for in-deploy COMM-s wherever there's a pair without X (a COMM)
+    //FIXME all `conflictsForNow` should eventually be replaced with `merges`
+    Map(
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
+      "!X !4"        -> conflictsForNow(S0, S0, F_),
+      "!X (!4)"      -> merges(S0, S1 | F_, Nil),
+      "!X !C"        -> conflictsForNow(S0, S1, C1),
+      "!X (!C)"      -> conflicts(S0, S1 | C_, Nil),
+      "!X 4X"        -> conflicts(S0, F_, Nil),
+      "!X 4!"        -> conflictsForNow(S0, F_, S0),
+      "!X (4!)"      -> coveredBy("!X (!4)"),
+      "!X 4!!"       -> conflictsForNow(S0, F_, R0),
+      "!X (4!!)"     -> conflictsForNow(S0, F_ | R0, Nil),
+      "!X !!X"       -> conflictsForNow(S0, R0, Nil),
+      "!X !!4"       -> conflictsForNow(S0, R1, F1),
+      "!X (!!4)"     -> coveredBy("!X (4!!)"),
+      "!X CX"        -> conflicts(S0, C_, Nil),
+      "!X C!"        -> conflicts(S0, C_, S0),
+      "!X (C!)"      -> conflicts(S0, C_ | S0, Nil),
+      "!4 !4 same 4" -> conflicts(S0, S1, F_),
+      "!4 !4 diff 4" -> conflicts(S0, S1, F0 | F1),
+      "!4 (!4)"      -> merges(S0, S1 | F1, F0),
+      "(!4) (!4)"    -> merges(S0 | F_, S0 | F_, Nil),
+      "!4 !C"        -> conflictsForNow(S0, S1, F0 | C1),
+      "!4 4X"        -> conflictsForNow(S0, F_, F_),
+      "!4 4!"        -> conflictsForNow(S0, F_, F0 | S1),
+      "!4 4!!"       -> conflictsForNow(S0, F_, F0 | R1),
+      "!4 !!X"       -> conflictsForNow(S0, R1, F0),
+      "!4 !!4"       -> conflictsForNow(S0, R1, F_ | F1),
+      "!4 CX"        -> conflictsForNow(S0, C_, F_),
+      "!4 C!"        -> conflictsForNow(S0, C_, F0 | S1),
+      "!C !C"        -> conflictsForNow(S0, S0, C_),
+      "!C 4X"        -> conflictsForNow(S0, F_, C_),
+      "!C 4!"        -> conflictsForNow(S0, F_, C0 | S1),
+      "!C 4!!"       -> conflictsForNow(S0, F_, C0 | R1),
+      "!C !!X"       -> conflictsForNow(S0, R1, C0),
+      "!C !!4"       -> conflictsForNow(S0, R1, C0 | F1),
```

Comment:
I think that it would be beneficial to have a `infinite` or `impossible` marker that could hold e.g. `4X C!!` for the sake of completeness.

Reply:
Shouldn't this `conflictForNow` and eventually `merge`? The impossible case is going to run out of phlo and so the deploy is going to be reverted without touching the tuplespace. So there shouldn't be any conflicts. Correct me if I'm wrong.

Reply:
depends on what the interpreter will do :D

Reply:
Future PR sounds good